src/diffusers/loaders/single_file_utils.py

```diff
@@ -1249,369 +1192,256 @@ def create_text_encoder_from_open_clip_checkpoint(
+            logger.warn(
+                f"Some weights of the model checkpoint were not used when initializing {cls.__name__}: \n {[', '.join(unexpected_keys)]}"
             )
 
     else:
-        if not (hasattr(text_model, "embeddings") and hasattr(text_model.embeddings.position_ids)):
-            text_model_dict.pop("text_model.embeddings.position_ids", None)
-
-        text_model.load_state_dict(text_model_dict)
+        model.load_state_dict(diffusers_format_checkpoint)
 
     if torch_dtype is not None:
-        text_model = text_model.to(torch_dtype)
+        model.to(torch_dtype)
 
-    return text_model
+    model.eval()
 
+    return model
 
-def create_diffusers_unet_model_from_ldm(
-    pipeline_class_name,
-    original_config,
+
+def create_diffusers_unet_from_ldm(
+    cls,
     checkpoint,
-    num_in_channels=None,
-    upcast_attention=None,
-    extract_ema=False,
-    image_size=None,
+    config=None,
+    original_config=None,
     torch_dtype=None,
-    model_type=None,
+    **kwargs,
 ):
-    from ..models import UNet2DConditionModel
+    num_in_channels = kwargs.pop("num_in_channels", None)
+    extract_ema = kwargs.pop("extract_ema", False)
 
-    if num_in_channels is None:
-        if pipeline_class_name in [
-            "StableDiffusionInpaintPipeline",
-            "StableDiffusionControlNetInpaintPipeline",
-            "StableDiffusionXLInpaintPipeline",
-            "StableDiffusionXLControlNetInpaintPipeline",
-        ]:
-            num_in_channels = 9
+    if original_config is not None:
+        model_type = kwargs.get("model_type", None)
+        model_type = infer_model_type(original_config, checkpoint, model_type=model_type)
 
-        elif pipeline_class_name == "StableDiffusionUpscalePipeline":
-            num_in_channels = 7
+        image_size = kwargs.get("image_size", None)
+        image_size = set_image_size(cls, original_config, checkpoint, image_size=image_size, model_type=model_type)
 
-        else:
-            num_in_channels = 4
+        model_config = create_unet_diffusers_config(original_config, image_size=image_size)
 
-    image_size = set_image_size(
-        pipeline_class_name, original_config, checkpoint, image_size=image_size, model_type=model_type
-    )
-    unet_config = create_unet_diffusers_config(original_config, image_size=image_size)
-    unet_config["in_channels"] = num_in_channels
-    if upcast_attention is not None:
-        unet_config["upcast_attention"] = upcast_attention
+    elif config is not None:
+        model_config = config
 
-    diffusers_format_unet_checkpoint = convert_ldm_unet_checkpoint(checkpoint, unet_config, extract_ema=extract_ema)
-    ctx = init_empty_weights if is_accelerate_available() else nullcontext
+    else:
+        config = fetch_diffusers_config(checkpoint, subfolder="unet")
+        model_config = cls.load_config(**config)
 
+    if num_in_channels is not None:
+        deprecate("num_in_channels", "in_channels", "1.0.0")
```

Comment:
We should deprecate with a message here.

Reply:
Also, I didn't get how are we going to take care of `num_in_channels` if it was None. Wouldn't be nice to have an `infer_num_in_channels` function to do this?